Pass connection rather than connection ID

eileencodes opened the pull request:
`connection_id` shouldn't be changed so we don't need to explicitly pass
it around. If you want to change the `connection_id` then `#with_connection`
should be used instead. Changing the `connection_id` directly could
result in issues with ActiveRecord connections.

This change allows the connection to be cached through the stack by
passing `connection` rather than `connection_id`.

I'm sending this as a PR instead of pushing it directly because @tenderlove and I are unsure about this change. Tests pass but you never know :wink:

matthewd commented on activerecord/lib/active_record/connection_handling.rb:
```diff
@@ -88,7 +88,7 @@ def connection
+      Thread.current.object_id
     end
 
     def connection_id=(connection_id)
```
Shouldn't this go away now?

tenderlove replied:
Ya, we should rm the setter.

eileencodes replied:
:+1: done

tenderlove commented:
/cc @jeremy

thedarkone commented:
On an unrelated note, in #14938 I'm removing `current_connection_id` from `ConnectionPool` :stuck_out_tongue:.

magikid commented:
Thanks for the pull request @eileencodes! It looks like since #14938 was merged into master, this isn't a problem anymore. Would you mind closing your pull request?

kaspth commented:
@magikid Just FYI @eileencodes is part of the Rails committers team and knows how this whole Pull Request thing works :smile:

Thanks though, I'll close :heart:

thedarkone commented:
In #14938 `connection_id` was removed from a different part of `ActiveRecord`, in my comment above I was pointing out that this PR is a step in right direction, since after this and #14938 lands, `connection_id` will be no more.

@kaspth please re-open.

kaspth commented:
Ah, shoot! Thanks @thedarkone :heart:

eileencodes committed:
Pass connection rather than connection ID

`connection_id` shouldn't be changed so we don't need to explicitly pass
it around. If you want to change the `connection_id` then `#with_connection`
should be used instead. Changing the `connection_id` directly could
result in issues with ActiveRecord connections.

This change allows the connection to be cached through the stack by
passing `connection` rather than `connection_id`.

eileencodes commented:
I finally got around to rebasing this :wink: Seems there are no objections to merging this?

rafaelfranca commented on activerecord/test/cases/query_cache_test.rb:
```diff
@@ -43,7 +43,7 @@ def test_exceptional_middleware_assigns_original_connection_id_on_error
 
     mw = ActiveRecord::QueryCache.new lambda { |env|
-      ActiveRecord::Base.connection_id = self.object_id
+      ActiveRecord::RuntimeRegistry.connection_id = self.object_id
```
Does this test still make sense?

arthurnn commented:
:shipit:

eileencodes commented:
Closing because this PR is old and 95% of the code this is editing is no longer there
